Make `push_error` and `push_warning` print name of calling function

TV4Fun opened the pull request:
Currently, `push_error` and `push_warning` give themselves as the calling
function, which is not useful for debugging. I have altered these to use the
C stacktrace to get the name of the function that called them. This saves
having to turn these functions into macros, which would require recompiling
every single piece of code that uses them. I have also set it to, when debug
symbols are available, use them with `atos` (available on macOS) to find the
exact source file and line being called from. It should be possible to the same
thing on Linux using `addr2line`, but I don't have a Linux box handy to test
that. I am not sure how you would implement this in Windows.

I have tested this on macOS Sonoma 14.5 (23F79). It should work on any *nix
system. I am not sure about Windows.

Fixes #76770

TV4Fun committed:
Allow constructing `List`s from `initializer_list`s

QOL improvement to allow direct construction of `List`s.

TV4Fun committed:
Make `push_error` and `push_warning` print name of calling function

Currently, `push_error` and `push_warning` give themselve as the calling
function, which is not useful for debugging. I have altered these to use the
C stacktrace to get the name of the function that called them. This saves
having to turn these functions into macros, which would require recompiling
every single piece of code that uses them. I have also set it to, when debug
symbols are available, use them with `atos` (available on macOS) to find the
exact source file and line being called from. It should be possible to the same
thing on Linux using `addr2line`, but I don't have a Linux box handy to test
that. I am not sure how you would implement this in Windows.

I have tested this on macOS Sonoma 14.5 (23F79). It should work on any *nix
system. I am not sure about Windows.

Fixes #76770

Mickeon commented:
Could you provide screenshots on how this would look now?

AThousandShips commented on core/templates/list.h:
```diff
@@ -761,6 +763,12 @@ class List {
 	}
 
+	List(std::initializer_list<T> init) {
+		for (const auto& i : init) {
```
```suggestion
		for (const T &i : init) {
```
`auto` is [not allowed](https://docs.godotengine.org/en/stable/contributing/development/cpp_usage_guidelines.html#auto-keyword)

TV4Fun replied:
Fixed.

AThousandShips reviewed (dismissed):
This needs to be done in a platform independent fashion, you need to fix the style issues so the CI can run to verify it on all platforms, but this would need to run on all platforms with reasonable support IMO, or be put just in the specific platform

AThousandShips commented:
Also I'm not sure this change makes sense, this method is called from scripting, the GDScript method won't be shown, so please add some screenshots to see what this outputs

TV4Fun committed:
Fix code style issues

TV4Fun committed:
Move C++ code line discovery to its own function

Also add method to print whole callstack.

TV4Fun committed:
Move platform dependent code into `OS` class and fix stacktrace from scripts

TV4Fun commented:
Okay, I have fixed the formatting and code style issues, moved the platform dependent code into relevant subclasses of `OS`, and fixed it so that it now properly reports the stack trace when called from scripts. This is only implemented for macOS for now as that is the system I have available. The Linux implementation should be similar except that retrieving debug symbols should be done with `addr2line` instead of `atos`. I am not sure exactly how to implement this on Windows, so perhaps someone with more knowledge of Windows debugging could contribute that part.

To clarify, there are two cases this fix covers:
1. **`push_error`/`push_warning` are called from GDScript code** This appears to have been partially implemented in the debugger for runtime code, but still fails for editor code. I have added code to retrieve the script callstack modeled after what `RemoteDebugger::_err_handler` does. In order for this to work requires that `DEBUG_ENABLED` be defined (true by default for `editor` and `template_debug` targets), and that Godot is run with the `--debug` flag.
2. **`push_error`/`push_warning` are called from C++ code** This can be from within the Godot engine or from GDExtension/GDNative code. This requires retrieving the C++ callstack. Getting source filenames and line numbers requires that the assembly generating the message was compiled with debug symbols. If debug symbols are not available, it can still usually get the name of the function that produced the message, which is much more useful than what is currently availble.

# Current behavior
## Runtime GDScript
As I said, this is partially implemented already, but still includes an unhelpful reference to `variant_utility.cpp` that is now fixed. Here is how it looked with a warning generated from a GDScript file before in debugger:

<img width="496" alt="Screenshot 2024-06-27 at 10 14 14 PM" src="https://github.com/godotengine/godot/assets/4702728/58c9eea1-2b03-40b3-a6c5-0195605f1f2c">

## Editor GDScript
GDScript run in the editor did not properly show where it was called from before, either in the editor logger or to stdout. Here is the view from the editor:

<img width="576" alt="Screenshot 2024-06-27 at 10 13 01 PM" src="https://github.com/godotengine/godot/assets/4702728/77059a6a-9864-49e9-97d4-52c27ff5e8cd">

and here is the view from stdout (with Rider):

<img width="534" alt="Screenshot 2024-06-27 at 10 13 08 PM" src="https://github.com/godotengine/godot/assets/4702728/cee385e1-820e-4334-a16e-ed76ef4e2bbd">

## C++
Here is a warning generated from `godot-git-plugin` in the editor logger:

<img width="543" alt="Screenshot 2024-06-27 at 10 13 35 PM" src="https://github.com/godotengine/godot/assets/4702728/a908ae8e-a571-4143-bb84-811e20d2a185">

and the view from stdout:

<img width="553" alt="Screenshot 2024-06-27 at 10 13 43 PM" src="https://github.com/godotengine/godot/assets/4702728/ddd35a46-d7d0-4be9-a0da-523cabc17eec">

# Proposed improvements
## Runtime GDScript

<img width="301" alt="Screenshot 2024-06-27 at 11 16 41 PM" src="https://github.com/godotengine/godot/assets/4702728/e64ede20-5a1f-49f3-b702-18d5aca6d448">

<img width="297" alt="Screenshot 2024-06-27 at 10 32 10 PM" src="https://github.com/godotengine/godot/assets/4702728/1516a987-c3a2-4302-a709-bfe445526b6d">

## Editor GDScript
View in editor logger:
<img width="421" alt="Screenshot 2024-06-27 at 10 16 50 PM" src="https://github.com/godotengine/godot/assets/4702728/5b5aa7e4-95d3-4f28-8c9f-9778505f743b">


View in stdout:

<img width="346" alt="Screenshot 2024-06-27 at 10 17 06 PM" src="https://github.com/godotengine/godot/assets/4702728/c25c3d31-5007-41ec-96ef-d38e13be669a">

## C++
View in editor logger:

<img width="778" alt="Screenshot 2024-06-27 at 10 17 39 PM" src="https://github.com/godotengine/godot/assets/4702728/21f0d3a4-fb1d-42c3-a2db-3f502d423984">

View in stdout:

<img width="1193" alt="Screenshot 2024-06-27 at 10 17 58 PM" src="https://github.com/godotengine/godot/assets/4702728/27de6c5c-e287-436f-9efe-7574eab5d959">

There are I'm sure still ways to make the results look a little better. All of the various logger implementations seem to have different ideas of what the two different string arguments to `errfunc` are for, which makes it hard to create a unified implementation that gives the relevant information without giving extra irrelevant information. I will happily accept any contributions anyone else wants to make to improve the output formatting.

TV4Fun committed:
Try fix for Linux builds

The Linux build breaks on `OS_Unix` because `Dl_info` is not `typedef`ed as
`struct dl_info`, so the forward declaration of that struct breaks. Changing
the forward declaration to `struct Dl_info` breaks the Mac build though, so
that leaves us with some `#ifdef`s. This may or may not work, I'll have to see.

TV4Fun committed:
Actual fix for Linux builds

Okay, there doesn't actually seem to be a way to forward declare a `typedef`
to an anonymous `struct`. So, rather than include an unnecessary system header
in `os_unix.h`, I am now just passing the members of `Dl_info` individually.
This will be slightly less efficient, but this code is generally not called
in performance-critical paths, and this should be clean, maintainable, and
platform-independent.

TV4Fun committed:
Fix Android and Web builds

Neither Android's NDK nor Emscripten provide `backtrace` through `execinfo.h`.
There are other ways to get a stacktrace on these platforms, but I am not
expert enough to implement them, so leave them unimplemented for now.

Calinou commented:
Would https://github.com/godotengine/godot/pull/91006 be of any use here?

TV4Fun commented:
> Would #91006 be of any use here?

Yes, that looks like it would do very much the same thing as what I do for GDScript stacktraces. Though theirs is a much more sweeping change, and mine is much more targeted. Mine requires being run in debug mode to get an accurate GDScript backtraces, while theirs completely rewrites GDScript's stack system to be fast enough to run all the time. Their are pros and cons to both approaches.

One thing my PR does that theirs doesn't is print the C++ stacktrace, which is still needed if `push_warning` or `push_error` are called from C++ code.

TV4Fun committed:
Remove unnecessary `i == 0` checks from string building loops

`String` is auto-initialized to be empty and `String::operator+=`
[already checks](https://github.com/godotengine/godot/blob/f0d15bbfdfde1c1076903afb7a7db373580d5534/core/string/ustring.cpp#L520)
if it is empty before appending, so this check is pointless and just needlessly
clutters the codebase.

AThousandShips commented on core/variant/variant_utility.cpp:
```diff
@@ -933,12 +933,7 @@ String VariantUtilityFunctions::str(const Variant **p_args, int p_arg_count, Cal
-		} else {
-			s += os;
-		}
+		s += os;
```
Why these unrelated changes? Please restore, if they should be done they should be done in a separate PR

TV4Fun replied:
Done.

AThousandShips commented on drivers/unix/os_unix.h:
```diff
@@ -101,6 +103,10 @@ class OS_Unix : public OS {
+
+#if !defined(__ANDROID__) && !defined(__EMSCRIPTEN__)
+	virtual Vector<StackInfo> get_cpp_stack_info() const override;
+#endif
```
These should be in the respective OS implementations I'd say, i.e. Mac and Linux, or overridden as stubs in Android and Web, this makes it hard to track

TV4Fun replied:
@AThousandShips I think it makes the most sense to keep them in `OS_Unix` because `execinfo.h` is theoretically a part of any standards-compliant Unix system. The problem is that Android and Emscripten aren't fully POSIX compliant and don't provide the `execinfo.h` header. I was surprised to even see that they were children of `OS_Unix` because neither are really Unix systems. I can remove the `#ifdef`s from the header here, but I will need to keep them in `os_unix.cpp` (which already has a ton of these because of the differences in different Unix implementations) because the implementation there depends on being able to import `execinfo.h`, which is not present in either

AThousandShips replied:
Fair enough, but this makes it harder to find as it'd be expected to be in those platforms instead of drivers

TV4Fun replied:
Yeah, it's weird to me that `OS_MacOS` in `platform` inherits from `OS_Unix` in `drivers`, which inherits from `OS` in `core`, and each provides different portions of the platform class's implementation, but here we are.

RandomShaper commented:
Precisely these days I'm trying to hunt down the root trigger of certain operations and the ability to obtain a proper C++ stack trace would come super handy.

Regarding the implementation, I'd just like to point out that the various `handle_crash()` implementations already contain code for stack trace collection. I'd say this PR would need to refactor those pieces out into individual functions instead of adding its own.

TV4Fun committed:
Revert "Remove unnecessary `i == 0` checks from string building loops"

This reverts commit 029246d8b6645a5f705f2ccd6beb5518f39aae21.

TV4Fun committed:
Prefix arguments with `p_`

Incorporate suggestions from @AThousandShips and necessary code changes to make
those work.

TV4Fun commented on core/templates/list.h:
```diff
@@ -761,6 +763,12 @@ class List {
 		}
 	}
 
+	List(std::initializer_list<T> init) {
```
Done.

TV4Fun commented on core/error/error_macros.h:
```diff
@@ -69,6 +69,8 @@ void _err_print_error(const char *p_function, const char *p_file, int p_line, co
 void _err_print_index_error(const char *p_function, const char *p_file, int p_line, int64_t p_index, int64_t p_size, const char *p_index_str, const char *p_size_str, const char *p_message = "", bool p_editor_notify = false, bool fatal = false);
 void _err_print_index_error(const char *p_function, const char *p_file, int p_line, int64_t p_index, int64_t p_size, const char *p_index_str, const char *p_size_str, const String &p_message, bool p_editor_notify = false, bool fatal = false);
+void _err_print_callstack(const String &p_error, bool p_editor_notify = false, ErrorHandlerType p_type = ERR_HANDLER_ERROR);
+void _err_print_error_backtrace(const char *filter, const String &p_error, bool p_editor_notify = false, ErrorHandlerType p_type = ERR_HANDLER_ERROR);
```
Done.

TV4Fun commented:
> Precisely these days I'm trying to hunt down the root trigger of certain operations and the ability to obtain a proper C++ stack trace would come super handy.
> 
> Regarding the implementation, I'd just like to point out that the various `handle_crash()` implementations already contain code for stack trace collection. I'd say this PR would need to refactor those pieces out into individual functions instead of adding its own.

@RandomShaper Yes, just looking at the MacOS implementation, there is definitely a lot of overlap in functionality that could be moved into a separate library to save some duplication. If you would like to contribute a refactor to do that, you are more than welcome to. Personally, I am happy with this version and need to move on to other projects.

bruvzg commented on platform/macos/os_macos.mm:
```diff
@@ -757,6 +757,19 @@
+#ifdef DEBUG_ENABLED
+String OS_MacOS::get_debug_descriptor(const StackInfo &p_info) {
+	String pipe;
+	Error error = execute("atos", { "-o", p_info.file, "-l", String::num_uint64(reinterpret_cast<uint64_t>(p_info.load_address), 16), String::num_uint64(reinterpret_cast<uint64_t>(p_info.symbol_address), 16) }, &pipe);
```
`atos` call probably should have `-arch arm64` or `-arch x86_64` arg added based on the running executable architecture, or in some conditions (like running under Rosetta) it might try reading wrong architecture from the fat binary.

RandomShaper commented:
> @RandomShaper Yes, just looking at the MacOS implementation, there is definitely a lot of overlap in functionality that could be moved into a separate library to save some duplication. If you would like to contribute a refactor to do that, you are more than welcome to. Personally, I am happy with this version and need to move on to other projects.

Would it be fine with you that someone (possibly me when I can spend time on this) takes over your PR? That would mean that other contributor would take your changes, apply the refactor and commit them to a new PR superseding this one, crediting both authors (by having separate commits if it makes sense for the repo or using a co-author annotation).

TV4Fun commented:
> > @RandomShaper Yes, just looking at the MacOS implementation, there is definitely a lot of overlap in functionality that could be moved into a separate library to save some duplication. If you would like to contribute a refactor to do that, you are more than welcome to. Personally, I am happy with this version and need to move on to other projects.
> 
> Would it be fine with you that someone (possibly me when I can spend time on this) takes over your PR? That would mean that other contributor would take your changes, apply the refactor and commit them to a new PR superseding this one, crediting both authors (by having separate commits if it makes sense for the repo or using a co-author annotation).

Yes, that would be fine. Just give me credit in some way.

reduz commented:
This is something I would really like to add to #91006 to make it more complete (my PR fixes a number of issues in GDScript and also makes it work properly outside the debugger, as well as making the log system support it properly too).

The problem is that It's kind of only usable for desktop, afaik on mobile and other platforms there is no implementation to get the C stack.
